### ISSUE TITLE
[BUG] Comment stripping for JSON in node 19 doesn't work

### ISSUE THREAD

**Issue by edomora97** [node-config/node-config#715]
**Describe the bug**

With Node 19 (not sure which is the minimum failing version) a json with comments (with .json extension) doesn't work.

**Expected behavior**
I expect that the comments are removed, as described by [the docs](https://github.com/node-config/node-config/wiki/Configuration-Files#file-formats).

**How to reproduce**

<details>
<summary>package.json</summary>

```json
{
  "name": "test",
  "version": "1.0.0",
  "main": "index.js",
  "dependencies": {
    "config": "^3.3.9"
  }
}
```

</details>

<details>
<summary>config/default.json</summary>

```json
{
    // Example comment
}
```

</details>

<details>
<summary>index.js</summary>

```js
const config = require("config");
```

</details>

Running `node index.js` it fails with:

```
/tmp/test/node_modules/config/lib/config.js:853
    throw new Error("Cannot parse config file: '" + fullFilename + "': " + e3);
    ^

Error: Cannot parse config file: '/tmp/test/config/default.json': SyntaxError: Expected property name or '}' in JSON at position 6
    at util.parseFile (/tmp/test/node_modules/config/lib/config.js:853:11)
    at /tmp/test/node_modules/config/lib/config.js:649:26
    at Array.forEach (<anonymous>)
    at util.loadFileConfigs (/tmp/test/node_modules/config/lib/config.js:648:16)
    at new Config (/tmp/test/node_modules/config/lib/config.js:113:27)
    at Object.<anonymous> (/tmp/test/node_modules/config/lib/config.js:1510:31)
    at Module._compile (node:internal/modules/cjs/loader:1239:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1293:10)
    at Module.load (node:internal/modules/cjs/loader:1096:32)
    at Module._load (node:internal/modules/cjs/loader:935:12)

Node.js v19.4.0
```

### Please tell us about your environment:

 -  node-config version: 3.3.9
 -  node-version: 19.4.0
 
### Other information

Probably the message changed (again?) and it fails to fallback to json5

https://github.com/node-config/node-config/blob/4c1c6193903bdb8e477792bec9dd852f7508dbcf/parser.js#L163

**Comment by markstos** [comment]
I'm not sure what changed here. PR welcome.

**Comment by TheGauntt** [comment]
@edomora97 I ran into the same issue. It looks like JSON.parse tries to be more helpful and has 3+ error conditions when trying to use comments in JSON files in Node 19.

One workaround is to use the `.json5` extension

**Comment by markstos** [comment]
I'd like to support comments in JSON with Node 19 if someone wants to figure out the patch.

**Comment by Eskibear** [comment]
Error message of `JSON.parse` has been changed.

`default.json`:
```
{
  /*comments*/
}
```

Node 16.14.2:
e.message = 'Unexpected token / in JSON at position 5'

Node 19.9.0:
e.message = 'Expected property name or '}' in JSON at position 5'

Culprit:
https://github.com/node-config/node-config/blob/46d0c31c4a14e38c115b3eb8807c880d9d1b5f0e/parser.js#L161-L163

**Comment by jfberry** [comment]
Is this also a problem with node 20? (I saw this with node 19 and rolled back)

**Comment by TheGauntt** [comment]
Yes

**Comment by jfberry** [comment]
We are blocked from moving to node 20 because of this change in behaviour. Is the PR going to be merged or other solution?

**Comment by ReuschelCGN** [comment]
Something new on this issu?

**Comment by markstos** [comment]
I just took a look at the history here. Here's what I found:

 * Prior to 1.18.0, we supported comments in JSON, but there was a bug, so it could mis-parsed. 
 * Discussion ensued in #256. The principled approach was discuss of leaving "JSON as JSON". With a breaking change in a major release, we would drop support for comments in JSON and tell folks to use JSON5 instead, which support that as part of the spec.
 * But there was pushback to keep backwards compatibility, resulting in the behavior we had today: We would parse JSON without stripping comments, but if the error message when it failed indicated that the problem was related to comment parsing, in that case we would fallback to JSON5 parsing. 

The proposal here is to /always/  fallback to JSON5 parsing if JSON parsing fails. This effectively means that all JSON5 features become valid in JSON fails, which was never the intent. 

This was something at the author, @lorenwest argued against then, and I agree that's not a great resolution.

Now, let's pivot to update from the JSON5 project to see what's happening with it:

 > JSON5 is an extension to the popular [JSON](https://tools.ietf.org/html/rfc7159) file format that aims to be easier to write and maintain by hand (e.g. for config files). It is not intended to be used for machine-to-machine communication. (Keep using JSON or other file formats for that. 🙂)

> JSON5 was started in 2012, and as of 2022, now gets [>65M downloads/week](https://www.npmjs.com/package/json5), ranks in the [top 0.1%](https://gist.github.com/anvaka/8e8fa57c7ee1350e3491) of the most depended-upon packages on npm, and has been adopted by major projects like [Chromium](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/runtime_enabled_features.json5;drc=5de823b36e68fd99009a29281b17bc3a1d6b329c), [Next.js](https://github.com/vercel/next.js/blob/b88f20c90bf4659b8ad5cb2a27956005eac2c7e8/packages/next/lib/find-config.ts#L43-L46), [Babel](https://babeljs.io/docs/en/config-files#supported-file-extensions), [Retool](https://community.retool.com/t/i-am-attempting-to-append-several-text-fields-to-a-google-sheet-but-receiving-a-json5-invalid-character-error/7626), [WebStorm](https://www.jetbrains.com/help/webstorm/json.html), and [more](https://github.com/json5/json5/wiki/In-the-Wild). It's also natively supported on [Apple platforms]
(https://developer.apple.com/documentation/foundation/jsondecoder/3766916-allowsjson5) like MacOS and iOS.

Rather than fade away, this JSON superset has gone on to to see wide adoption and endorsement from major projects, _precisely because it's good for cases like ours--confg files_

So at this point, I propose instead of this approach, we completely drop the use of the regular JSON parser, and always use the JSON5 parser. This would be backwards compatible for all users, and is a rare case where we can both add a feature and simplify the code.

If @lorenwest doesn't comment in >1 week, I'll proceed with that plan. In the meantime, if someone wants to move the proposal along, they could add a PR for it, and make sure we have test coverage. Our docs should be updated now and the wiki after the release.

**Comment by markstos** [comment]
Anyone who needs an immediate workaround and rename their JSON config files to `.json5` and make sure to the JSON5 module installed. That will trigger our JSON5 parser code path, where comments in JSON have always been supported.

**Comment by lorenwest** [comment]
I'm ok with that solution.  Doubt many would care about the extra strictness of JSON vs. JSON5, although it wouldn't surprise me if someone commented about it 😄

**Comment by markstos** [comment]
Thanks. PR is welcome then to implement the above.

**Comment by markstos** [comment]
Fixing the test failure described here would also be welcome:

https://github.com/node-config/node-config/issues/704#issuecomment-1529759496

**Comment by jfberry** [comment]
This issue also affects bun runtime as well as node 19+

**Comment by jaggehns** [comment]
> This issue also affects bun runtime as well as node 19+

When using ts files such as default.ts in an Express application, the error below occurs when running `bun build/src/app.js`

[0.41ms] ".env"
848 |   }
849 |   catch (e3) {
850 |     if (gitCryptTestRegex.test(fileContent)) {
851 |       console.error('ERROR: ' + fullFilename + ' is a git-crypt file and CONFIG_SKIP_GITCRYPT is not set.');
852 |     }
853 |     throw new Error("Cannot parse config file: '" + fullFilename + "': " + e3);
              ^
error: Cannot parse config file: '$PATH/config/default.ts': TypeError: module_1.Module._preloadModules is not a function. (In 'module_1.Module._preloadModules(service.options.require)', 'module_1.Module._preloadModules' is undefined)

     at /node_modules/config/lib/config.js:853:10
      at /node_modules/config/lib/config.js:649:20
      at forEach (:1:20)
      at /node_modules/config/lib/config.js:648:2
      at new Config (/node_modules/config/lib/config.js:113:21)
      at /node_modules/config/lib/config.js:1510:30
      at globalThis (/node_modules/config/lib/config.js:1520:101)
      at require (:1:20)
      at /build/src/app.js:7:17
      at globalThis (/build/src/app.js:14:1)
      
      This can be recreated in any Express app that uses node-config and the Bun environment

**Comment by DeutscherDude** [comment]
Hello :smile: ,
I have been looking into both the parsing issue & #704.
As for the JSON parsing error for invalid token '}', it seemed to be the matter of a RegExp issue. It was previously matching
![Failing-Regexp](https://github.com/node-config/node-config/assets/86159966/0e184fed-26e5-48f0-b4ad-4da7a4e720b5)

So I changed it to the following and the tests started passing again.
```js
  stringRegex = stringRegex || /(['"])(\\\1|[a-z0-9 _\/\{\}\?\=\<\>\:\.]|)+?\1/gi;
```
![FixedRegexp](https://github.com/node-config/node-config/assets/86159966/3eeefd8b-7d1f-4956-be00-66d3478d758d)

I will continue investigating the #704 error, so far nothing groundbreaking, the only thing I noticed is the actual object returned from `CONFIG.get('aPromise')` is a proxied Promise.
![RawPromiseReturn](https://github.com/node-config/node-config/assets/86159966/721f821a-0cb5-4e3c-b138-3b2f9103c3d1)

**Comment by lorenwest** [comment]
@DeutscherDude I see your proposed solution is to change the regexp vs. the above suggestion of fallback to JSON5. Is there a reason you feel this is better than JSON5 parsing for JSON?

**Comment by DeutscherDude** [comment]
@lorenwest nothing in particular drives me towards correcting the RegExp over using json5 parser.
The main thing in my mind whilst introducing the update to RegExp was consistency with previous iterations of the code and keeping it predictable, i.e. using the parser dedicated to json rather than json5.
Overall, I am happy to test the json5 option and update my PR 😊

**Comment by lorenwest** [comment]
I'm on the fence as well.  Anyone else want to chime in with thoughts beyond the "i don't like supporting regexp" vs. "i don't like changing too much" issues described above?

**Comment by lorenwest** [comment]
It seems that regexps are forever riddled with edge case problems, especially ones with [a-zA-Z] in multi-language settings.  For that reason I lean toward having someone else deal with it, and moving to JSON5 for all JSON parsing.

**Comment by DeutscherDude** [comment]
Understood, will amend my PR accordingly 😊

**Comment by jfberry** [comment]
> Understood, will amend my PR accordingly 😊

any progress on this?

**Comment by PanMan** [comment]
This issue is still open but it does seem to be fixed in the latest version, `3.3.11` But didn't see it in the release notes?
https://github.com/node-config/node-config/commit/00587019e2013656f470de925d5187929583cbec

**Comment by lorenwest** [comment]
This issue is closed, as the default parser has been changed to the JSON5 parser.
